cuda_core/cuda/core/experimental/_stream.pyx

```diff
@@ -168,46 +165,37 @@ cdef class Stream:
             self._owner = obj
             self._nonblocking = None  # delayed
             self._priority = None  # delayed
-            self._device_id = None  # delayed
-            self._ctx_handle = None  # delayed
             return self
 
         cdef StreamOptions opts = check_or_create_options(StreamOptions, options, "Stream options")
         nonblocking = opts.nonblocking
         priority = opts.priority
 
         flags = cydriver.CUstream_flags.CU_STREAM_NON_BLOCKING if nonblocking else cydriver.CUstream_flags.CU_STREAM_DEFAULT
+        # TODO: we might want to consider memoizing high/low per CUDA context and avoid this call
         cdef int high, low
-        HANDLE_RETURN(cydriver.cuCtxGetStreamPriorityRange(&high, &low))
+        with nogil:
+            HANDLE_RETURN(cydriver.cuCtxGetStreamPriorityRange(&high, &low))
+        cdef int prio
         if priority is not None:
-            if not (low <= priority <= high):
+            prio = priority
+            if not (low <= prio <= high):
                 raise ValueError(f"{priority=} is out of range {[low, high]}")
         else:
-            priority = high
+            prio = high
 
         cdef cydriver.CUstream s
-        HANDLE_RETURN(cydriver.cuStreamCreateWithPriority(&s, flags, priority))
+        with nogil:
+            HANDLE_RETURN(cydriver.cuStreamCreateWithPriority(&s, flags, prio))
         self._handle = s
         self._owner = None
         self._nonblocking = nonblocking
         self._priority = priority
-        self._device_id = device_id
-        self._ctx_handle = None  # delayed
+        self._device_id = device_id if device_id is not None else self._device_id
         return self
 
-    def __del__(self):
-        self._shutdown_safe_close()
-
-    cdef _shutdown_safe_close(self, is_shutting_down=sys.is_finalizing):
-        if is_shutting_down and is_shutting_down():
-            return
-
-        if self._owner is None:
-            if self._handle and not self._builtin:
-                HANDLE_RETURN(cydriver.cuStreamDestroy(self._handle))
-        else:
-            self._owner = None
-        self._handle = <cydriver.CUstream>(NULL)
+    def __dealloc__(self):
+        self.close()
```

Comment:
Note: With @brandon-b-miller's reproducer (https://github.com/NVIDIA/cuda-python/issues/803#issuecomment-3292781448), I confirm this change continues to avoid shutdown issues as expected, although it'd be nice to have more ways to test this...